File: snail/manchann.py
#include <iostream>
using namespace std;

int n,m;
double cache[1001][2001] = {-1.0,};

double snail(int count, int height){
    double &ret = cache[count][height];


    if(ret != -1.0){
        return ret;
    }
    if(count == m){
        return height >= n;
    }

    return ret = (0.25 * snail(count+1,height+1)) + (0.75 * snail(count+1,height+2));
}

int main(){
    int cases;
    cin >> cases;

    while(cases--){
        cin>> n >> m;
        for(int i=0; i<1001; i++){
            for(int j=0; j<2001; j++ ){
                cache[i][j] = -1.0;
            }
        }
        cout.precision(11);
        cout<< snail(0,0) <<"\n";
    }

}

File: ASYMTILING/manchann.cpp
#include <iostream>
using namespace std;

int main(){
    int cases;
    cin>> cases;

    while(cases--){
        long long dp[101] = {0,};
        dp[1] = 1;
        dp[2] = 2;
    
        int num;
        cin>>num;
        if(num == 1){
            cout<<1<<"\n";
            continue;
        } else if(num == 2){
            cout<<0<<"\n";
            continue;
        }
        for(int i=3; i<=num; i++){
            dp[i] = (dp[i-1] + dp[i-2]) % 1000000007;
        }
        int ret =0;
        if(num %2 ==0){
            ret =1000000007+ dp[num] - ( dp[num/2] + dp[num/2 -1])%1000000007;
        } else{
            ret = 1000000007+dp[num] - dp[(num-1)/2]%1000000007;
        }
        cout<<ret % 1000000007<<"\n";
    }
}
